elftools/dwarf/callframe.py

```diff
@@ -193,6 +231,171 @@ def _parse_instructions(self, structs, offset, end_offset):
             offset = self.stream.tell()
         return instructions
 
+    def _parse_cie_for_fde(self, fde_offset, fde_header, entry_structs):
+        """ Parse the CIE that corresponds to an FDE.
+        """
+        # Determine the offset of the CIE that corresponds to this FDE
+        if self.for_eh_frame:
+            # CIE_pointer contains the offset for a reverse displacement from
+            # the section offset of the CIE_pointer field itself (not from the
+            # FDE header offset).
+            cie_displacement = fde_header['CIE_pointer']
+            cie_offset = (fde_offset + entry_structs.dwarf_format // 8
+                          - cie_displacement)
+        else:
+            cie_offset = fde_header['CIE_pointer']
+
+        # Then read it
+        with preserve_stream_pos(self.stream):
+            return self._parse_entry_at(cie_offset)
+
+    def _parse_cie_augmentation(self, header, entry_structs):
+        """ Parse CIE augmentation data from the annotation string in `header`.
+
+        Return a tuple that contains 1) the augmentation data as a string
+        (without the length field) and 2) the augmentation data as a dict.
+        """
+        augmentation = header.get('augmentation')
+        if not augmentation:
+            return ('', {})
+
+        # Augmentation parsing works in minimal mode here: we need the length
+        # field to be able to skip unhandled augmentation fields.
+        assert augmentation.startswith(b'z'), (
+            'Unhandled augmentation string: {}'.format(repr(augmentation)))
+
+        available_fields = {
+            b'z': entry_structs.Dwarf_uleb128('length'),
+            b'L': entry_structs.Dwarf_uint8('LSDA_encoding'),
+            b'R': entry_structs.Dwarf_uint8('FDE_encoding'),
+            b'S': True,
+            b'P': Struct(
+                'personality',
+                entry_structs.Dwarf_uint8('encoding'),
+                Switch('function', lambda ctx: ctx.encoding & 0x0f, {
+                    enc: fld_cons('function')
+                    for enc, fld_cons
+                    in self._eh_encoding_to_field(entry_structs).items()})),
+        }
+
+        # Build the Struct we will be using to parse the augmentation data.
+        # Stop as soon as we are not able to match the augmentation string.
+        fields = []
+        aug_dict = {}
+        for i in range(len(augmentation)):
+            try:
+                fld = available_fields[augmentation[i:i+1]]
+            except KeyError:
+                break
+
+            if fld is True:
+                aug_dict[fld] = True
+            else:
+                fields.append(fld)
+
+        # Read the augmentation twice: once with the Struct, once for the raw
+        # bytes. Read the raw bytes last so we are sure we leave the stream
+        # pointing right after the augmentation.
```

Comment:
Can the comment further explain this logic? Won't reading with Struct leave it in the right place? You need both anyway, right?

Reply:
First, we indeed need both the bytes for readelf dump and the parsed struct to extract information (such as the `FDE_encoding` field). Besides, if the augmentation is not totally supported (unknown char in it), then the Struct will be incomplete: trailing fields will be missing and thus we need to read raw bytes to get past the augmentation data. I will add these details to the comment.